File: lib/server/handlers/storage.ts
import * as db from "../db";
import { ensureUnlocked, folderTokenFromRequest } from "../locks";
import { classify, contentTypeFor, extOf, isMedia } from "../media";
import * as r2 from "../r2";
import { requireUser } from "../security";
import {
  assertOwned,
  exposeListing,
  stripUserRoot,
  toStoragePrefix,
} from "../user-scope";
import { visualIndexImage } from "../visual-index";

async function decorate(userId: string, items: Array<Record<string, unknown>>) {
  const locked = new Set(await db.listLockedFolders(userId));
  for (const it of items) {
    const storageKey = toStoragePrefix(userId, String(it.key));
    if (it.type === "folder") {
      it.locked = locked.has(storageKey);
    } else {
      const row = await db.getItem(userId, storageKey);
      it.favorite = Boolean(row?.favorite);
      it.tags = row?.tags ? row.tags.split(",").filter(Boolean) : [];
    }
  }
  return items;
}

export async function handleList(req: Request) {
  const user = await requireUser(req);
  const ft = folderTokenFromRequest(req);
  const { searchParams } = new URL(req.url);
  const prefix = searchParams.get("prefix") || "";
  const storagePrefix = toStoragePrefix(user.userId, prefix);
  await ensureUnlocked(storagePrefix, ft, user.userId);
  const result = await r2.listDir(storagePrefix);
  const exposed = exposeListing(result, user.userId);
  exposed.folders = (await decorate(user.userId, exposed.folders as Array<Record<string, unknown>>)) as typeof exposed.folders;
  exposed.files = (await decorate(user.userId, exposed.files as Array<Record<string, unknown>>)) as typeof exposed.files;
  return exposed;
}

export async function handleCreateFolder(req: Request) {
  const user = await requireUser(req);
  const body = await req.json();
  const storagePrefix = toStoragePrefix(user.userId, body.prefix || "");
  const key = await r2.createFolder(storagePrefix, body.name);
  return { key: stripUserRoot(user.userId, key) };
}

export async function handleUploadUrl(req: Request) {
  const user = await requireUser(req);
  const ft = folderTokenFromRequest(req);
  const body = await req.json();
  const prefix = toStoragePrefix(user.userId, body.prefix || "");
  await ensureUnlocked(prefix, ft, user.userId);
  const finalName =
    typeof body.relative_path === "string" && body.relative_path.trim()
      ? r2.sanitizeRelativePath(body.relative_path)
      : r2.sanitizeSegment(body.name);
  if (!isMedia(finalName.split("/").pop() || finalName)) {
    throw new r2.StorageError("Unsupported file type. Allowed: images, video and audio formats.");
  }
  const key = `${prefix}${finalName}`;
  const content_type = body.content_type || contentTypeFor(finalName);
  const url = await r2.presignPut(key);
  return { key: stripUserRoot(user.userId, key), url, content_type };
}

export async function handleUpload(req: Request) {
  const user = await requireUser(req);
  const ft = folderTokenFromRequest(req);
  const form = await req.formData();
  const prefix = String(form.get("prefix") || "");
  const relativePath = String(form.get("relative_path") || "");
  const file = form.get("file");
  if (!(file instanceof File)) throw new r2.StorageError("No file provided.");
  const storagePrefix = toStoragePrefix(user.userId, prefix);
  await ensureUnlocked(storagePrefix, ft, user.userId);
  const fileName = relativePath
    ? r2.sanitizeRelativePath(relativePath)
    : r2.sanitizeSegment(file.name || "upload");
  if (!isMedia(fileName.split("/").pop() || fileName)) {
    throw new r2.StorageError("Unsupported file type. Allowed: images, video and audio formats.");
  }
  const key = `${storagePrefix}${fileName}`;
  const name = fileName.split("/").pop() || fileName;
  const content_type = file.type || contentTypeFor(name);
  const buf = Buffer.from(await file.arrayBuffer());
  await r2.putObject(key, buf, content_type);
  await db.upsertItem(user.userId, {
    key,
    name,
    folder: r2.folderOf(key),
    type: classify(name),
    ext: extOf(name),
    size: buf.length,
    content_type,
    last_modified: null,
  });
  if (classify(name) === "image") {
    void visualIndexImage(user.userId, key, buf.length).catch(() => {});
  }
  return { key: stripUserRoot(user.userId, key) };
}

export async function handleIndexOne(req: Request) {
  const user = await requireUser(req);
  const ft = folderTokenFromRequest(req);
  const body = await req.json();
  const key = assertOwned(user.userId, body.key);
  await ensureUnlocked(key, ft, user.userId);
  const meta = await r2.headObject(key);
  if (!meta) throw new r2.StorageError("Upload not found in storage.");
  const name = key.split("/").pop() || "";
  await db.upsertItem(user.userId, {
    key,
    name,
    folder: r2.folderOf(key),
    type: classify(name),
    ext: extOf(name),
    size: meta.size,
    content_type: meta.content_type || contentTypeFor(name),
    last_modified: meta.last_modified,
  });
  let visual: { indexed: boolean; tags: string[] } | undefined;
  if (classify(name) === "image") {
    visual = await visualIndexImage(user.userId, key, meta.size);
  }
  return { ok: true, visual_tags: visual?.tags };
}

export async function handleMediaUrl(req: Request) {
  const user = await requireUser(req);
  const ft = folderTokenFromRequest(req);
  const { searchParams } = new URL(req.url);
  const key = assertOwned(user.userId, searchParams.get("key") || "");
  const download = searchParams.get("download") === "true";
  await ensureUnlocked(key, ft, user.userId);
  return { url: await r2.presignGet(key, download) };
}

export async function handleDelete(req: Request) {
  const user = await requireUser(req);
  const ft = folderTokenFromRequest(req);
  const body = await req.json();
  const storageKey = assertOwned(user.userId, body.key);
  await ensureUnlocked(storageKey, ft, user.userId);
  const deleted = await r2.deleteKey(storageKey);
  if (storageKey.endsWith("/")) {
    await db.deleteItemsUnder(user.userId, storageKey);
    await db.removeFolderLock(user.userId, storageKey);
  } else {
    await db.deleteItem(user.userId, storageKey);
  }
  return { deleted: deleted.map((k) => stripUserRoot(user.userId, k)) };
}

export async function handleMove(req: Request) {
  const user = await requireUser(req);
  const ft = folderTokenFromRequest(req);
  const body = await req.json();
  const src = assertOwned(user.userId, body.source);
  const isFolder = src.endsWith("/");
  const base = isFolder ? src.replace(/\/$/, "").split("/").pop()! : src.split("/").pop()!;
  const name = body.new_name ? r2.sanitizeSegment(body.new_name) : base;
  const destPrefix = toStoragePrefix(user.userId, body.destination_prefix || "");
  const dst = `${destPrefix}${name}${isFolder ? "/" : ""}`;
  await ensureUnlocked(src, ft, user.userId);
  await ensureUnlocked(dst, ft, user.userId);
  const newKey = await r2.moveKey(src, dst);
  if (isFolder) {
    await db.renamePrefix(user.userId, src, newKey);
  } else {
    const old = await db.getItem(user.userId, src);
    await db.deleteItem(user.userId, src);
    await db.upsertItem(user.userId, {
      key: newKey,
      name: newKey.split("/").pop() || "",
      folder: r2.folderOf(newKey),
      type: classify(newKey),
      ext: extOf(newKey),
      size: old?.size ?? null,
      content_type: contentTypeFor(newKey),
      last_modified: old?.last_modified ?? null,
    });
  }
  return { key: stripUserRoot(user.userId, newKey) };
}

export async function handleFavorite(req: Request) {
  const user = await requireUser(req);
  const body = await req.json();
  const storageKey = assertOwned(user.userId, body.key);
  await db.setFavorite(user.userId, storageKey, Boolean(body.favorite));
  return { ok: true };
}

File: backend/app/routers/storage.py
"""Storage browsing + mutation endpoints (all require a valid session)."""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, File, Form, UploadFile

from .. import db, r2
from ..locks import ensure_unlocked, folder_token_header
from ..media import classify, content_type_for, ext_of, is_media
from ..schemas import (
    CreateFolderRequest,
    DeleteRequest,
    FavoriteRequest,
    MoveRequest,
    PresignUploadRequest,
    PresignUploadResponse,
)
from ..security import CurrentUser
from ..user_scope import assert_owned, expose_listing, strip_user_root, to_storage_prefix

router = APIRouter(prefix="/api/storage", tags=["storage"], dependencies=[])

FolderToken = Annotated[str | None, Depends(folder_token_header)]


def _decorate(user_id: str, items: list[dict]) -> list[dict]:
    """Merge index metadata (favorite/tags) + lock flags into listing items."""
    locked = set(db.list_locked_folders(user_id))
    for it in items:
        storage_key = to_storage_prefix(user_id, it["key"])
        if it["type"] == "folder":
            it["locked"] = storage_key in locked
        else:
            row = db.get_item(user_id, storage_key)
            it["favorite"] = bool(row["favorite"]) if row else False
            it["tags"] = (
                [t for t in row["tags"].split(",") if t] if row and row["tags"] else []
            )
    return items


@router.get("/list")
async def list_dir(
    user: CurrentUser,
    folder_token: FolderToken,
    prefix: str = "",
) -> dict:
    storage_prefix = to_storage_prefix(user.user_id, prefix)
    ensure_unlocked(storage_prefix, folder_token, user.user_id)
    result = r2.list_dir(storage_prefix)
    exposed = expose_listing(result, user.user_id)
    exposed["folders"] = _decorate(user.user_id, exposed["folders"])
    exposed["files"] = _decorate(user.user_id, exposed["files"])
    return exposed


@router.post("/folder")
async def create_folder(user: CurrentUser, body: CreateFolderRequest) -> dict:
    storage_prefix = to_storage_prefix(user.user_id, body.prefix)
    key = r2.create_folder(storage_prefix, body.name)
    return {"key": strip_user_root(user.user_id, key)}


@router.post("/upload-url", response_model=PresignUploadResponse)
async def upload_url(
    user: CurrentUser,
    folder_token: FolderToken,
    body: PresignUploadRequest,
) -> PresignUploadResponse:
    prefix = to_storage_prefix(user.user_id, body.prefix)
    ensure_unlocked(prefix, folder_token, user.user_id)
    name = r2.sanitize_segment(body.name)
    if not is_media(name):
        raise r2.StorageError(
            "Unsupported file type. Allowed: images, video and audio formats."
        )
    key = f"{prefix}{name}"
    content_type = body.content_type or content_type_for(name)
    url = r2.presign_put(key, content_type)
    return PresignUploadResponse(
        key=strip_user_root(user.user_id, key),
        url=url,
        content_type=content_type,
    )


@router.post("/upload")
async def upload_file(
    user: CurrentUser,
    folder_token: FolderToken,
    prefix: str = Form(""),
    file: UploadFile = File(...),
) -> dict:
    """Server-side upload fallback when direct browser → R2 PUT is blocked (e.g. CORS)."""
    name = r2.sanitize_segment(file.filename or "upload")
    if not is_media(name):
        raise r2.StorageError(
            "Unsupported file type. Allowed: images, video and audio formats."
        )
    storage_prefix = to_storage_prefix(user.user_id, prefix)
    ensure_unlocked(storage_prefix, folder_token, user.user_id)
    key = f"{storage_prefix}{name}"
    content_type = file.content_type or content_type_for(name)
    body = await file.read()
    r2.put_object(key, body, content_type)
    db.upsert_item(
        user.user_id,
        {
            "key": key,
            "name": name,
            "folder": r2.folder_of(key),
            "type": classify(name),
            "ext": ext_of(name),
            "size": len(body),
            "content_type": content_type,
            "last_modified": None,
        },
    )
    return {"key": strip_user_root(user.user_id, key)}


@router.post("/index-one")
async def index_one(
    user: CurrentUser,
    folder_token: FolderToken,
    body: DeleteRequest,
) -> dict:
    """Record an object in the local index after a direct-to-R2 upload."""
    key = assert_owned(user.user_id, body.key)
    ensure_unlocked(key, folder_token, user.user_id)
    if not r2.object_exists(key):
        raise r2.StorageError("Upload not found in storage.")
    name = key.split("/")[-1]
    db.upsert_item(
        user.user_id,
        {
            "key": key,
            "name": name,
            "folder": r2.folder_of(key),
            "type": classify(name),
            "ext": ext_of(name),
            "size": None,
            "content_type": content_type_for(name),
            "last_modified": None,
        },
    )
    return {"ok": True}


@router.get("/url")
async def media_url(
    user: CurrentUser,
    folder_token: FolderToken,
    key: str,
    download: bool = False,
) -> dict:
    storage_key = assert_owned(user.user_id, key)
    ensure_unlocked(storage_key, folder_token, user.user_id)
    return {"url": r2.presign_get(storage_key, download=download)}


@router.post("/delete")
async def delete(
    user: CurrentUser,
    folder_token: FolderToken,
    body: DeleteRequest,
) -> dict:
    storage_key = assert_owned(user.user_id, body.key)
    ensure_unlocked(storage_key, folder_token, user.user_id)
    deleted = r2.delete_key(storage_key)
    if storage_key.endswith("/"):
        db.delete_items_under(user.user_id, storage_key)
        db.remove_folder_lock(user.user_id, storage_key)
    else:
        db.delete_item(user.user_id, storage_key)
    return {"deleted": [strip_user_root(user.user_id, k) for k in deleted]}


@router.post("/move")
async def move(
    user: CurrentUser,
    folder_token: FolderToken,
    body: MoveRequest,
) -> dict:
    src = assert_owned(user.user_id, body.source)
    is_folder = src.endswith("/")
    base = (src.rstrip("/").split("/")[-1]) if is_folder else src.split("/")[-1]
    name = r2.sanitize_segment(body.new_name) if body.new_name else base
    dest_prefix = to_storage_prefix(user.user_id, body.destination_prefix)
    dst = f"{dest_prefix}{name}{'/' if is_folder else ''}"

    ensure_unlocked(src, folder_token, user.user_id)
    ensure_unlocked(dst, folder_token, user.user_id)

    new_key = r2.move_key(src, dst)
    if is_folder:
        db.rename_prefix(user.user_id, src, new_key)
    else:
        old = db.get_item(user.user_id, src)
        db.delete_item(user.user_id, src)
        db.upsert_item(
            user.user_id,
            {
                "key": new_key,
                "name": new_key.split("/")[-1],
                "folder": r2.folder_of(new_key),
                "type": classify(new_key),
                "ext": ext_of(new_key),
                "size": old["size"] if old else None,
                "content_type": content_type_for(new_key),
                "last_modified": old["last_modified"] if old else None,
            },
        )
    return {"key": strip_user_root(user.user_id, new_key)}


@router.post("/favorite")
async def favorite(user: CurrentUser, body: FavoriteRequest) -> dict:
    storage_key = assert_owned(user.user_id, body.key)
    db.set_favorite(user.user_id, storage_key, body.favorite)
    return {"ok": True}
